arch/xtensa/core/userspace.S

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2022, Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <xtensa-asm2-s.h>
+#include <offsets.h>
+#include <offsets_short.h>
+#include <zephyr/syscall.h>
+#include <zsr.h>
+
+/**
+ *  syscall number     arg1, arg2, arg3, arg4, arg5, arg6
+ *  --------------     ----------------------------------
+ *  a8                 a2,   a3,   a4,   a5,   a6,   a7
+ *
+ **/
+.pushsection .text.z_xtensa_do_syscall, "ax"
+.global	z_xtensa_do_syscall
+.align	4
+z_xtensa_do_syscall:
+	rsr a0, ZSR_CPU
+	l32i a0, a0, ___cpu_t_current_OFFSET
+	l32i a0, a0, _thread_offset_to_psp
+
+	addi a0, a0, -___xtensa_irq_bsa_t_SIZEOF
+
+	s32i a1, a0, ___xtensa_irq_bsa_t_scratch_OFFSET
+	s32i a2, a0, ___xtensa_irq_bsa_t_a2_OFFSET
+	s32i a3, a0, ___xtensa_irq_bsa_t_a3_OFFSET
+	rsr a2, ZSR_ALLOCA
+	s32i a2, a0, ___xtensa_irq_bsa_t_a0_OFFSET
+	rsr.ps a2
+	s32i a2, a0, ___xtensa_irq_bsa_t_ps_OFFSET
+	rsr.epc1 a2
+	s32i a2, a0, ___xtensa_irq_bsa_t_pc_OFFSET
+
+	movi a2, PS_WOE|PS_INTLEVEL(XCHAL_NMILEVEL)
+	rsr.ps a3
+	or a3, a3, a2
+	movi a2, ~(PS_EXCM | PS_RING(3))
+	and a3, a3, a2
+	wsr.ps a3
+	rsync
+	l32i a2, a0, ___xtensa_irq_bsa_t_a2_OFFSET
+	l32i a3, a0, ___xtensa_irq_bsa_t_a3_OFFSET
+	SPILL_ALL_WINDOWS
+
```

Comment:
Coming back to clarify: this bit right here just can't merge.  We can't do ABI-defined window spills on behalf of unprivileged code, period.[1]  Now, it's 100% true that that's not really the fault of this PR, our existing entry code has always done it this way because the author didn't think about how that was going to translate to privilege boundaries, and that's on me and not you.  But still, there's just no way to salvage this particular entry path; whatever we use can't look like this.  We can't merge it.

[1] The spill has to be cooked through an interpretation layer of some form that prevents following the untrusted pointers.  Design note digression: In practice I'm basically convinced that the only solid way to do this is to unconditionally spill all 32/64 registers, maybe with some WINDOWSTART logic to elide quads that aren't showing as part of ABI frames (though that optimization would then break "creative" userspace users that want to do non-ABI stuff like green threads or coroutines, etc...).  But the tests are almost as expensive as the spills, so not sure that's worth it.  Though there are two ways to do this: interrupts need to do the spill in kernel mode, but syscalls can be optimized to do it in userspace before entering the kernel, which can be done with exceptions and thus automatically elide spills for unused frames.  Then the kernel can just clobber the registers knowing that userspace promised it doesn't care.  And of course, on the way OUT of the kernel we need to write them all with something (doesn't matter what) to avoid leaking stale kernel-stored values in those registers.